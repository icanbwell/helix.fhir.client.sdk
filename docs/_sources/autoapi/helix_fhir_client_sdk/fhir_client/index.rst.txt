:py:mod:`helix_fhir_client_sdk.fhir_client`
===========================================

.. py:module:: helix_fhir_client_sdk.fhir_client


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   helix_fhir_client_sdk.fhir_client.FhirClient




Attributes
~~~~~~~~~~

.. autoapisummary::

   helix_fhir_client_sdk.fhir_client.HandleBatchFunction
   helix_fhir_client_sdk.fhir_client.HandleErrorFunction


.. py:data:: HandleBatchFunction
   

   

.. py:data:: HandleErrorFunction
   

   

.. py:class:: FhirClient

   Class used to call FHIR server (uses async and parallel execution to speed up)

   .. py:method:: action(self, action)

      :param action: (Optional) do an action e.g., $everything


   .. py:method:: action_payload(self, action_payload)

      :param action_payload: (Optional) if action such as $graph needs a http payload


   .. py:method:: resource(self, resource)

      :param resource: what FHIR resource to retrieve


   .. py:method:: id_(self, id_)


   .. py:method:: url(self, url)

      :param url: server to call for FHIR


   .. py:method:: validation_server_url(self, validation_server_url)

      :param validation_server_url: server to call for FHIR validation


   .. py:method:: additional_parameters(self, additional_parameters)

      :param additional_parameters: Any additional parameters to send with request


   .. py:method:: filter_by_resource(self, filter_by_resource)

      :param filter_by_resource: filter the resource by this. e.g., /Condition?Patient=1
              (resource=Condition, filter_by_resource=Patient)


   .. py:method:: filter_parameter(self, filter_parameter)

      :param filter_parameter: Instead of requesting ?patient=1,
              do ?subject:Patient=1 (if filter_parameter is subject)


   .. py:method:: include_only_properties(self, include_only_properties)

      :param include_only_properties: includes only these properties


   .. py:method:: page_number(self, page_number)

      :param page_number: page number to load


   .. py:method:: page_size(self, page_size)

      :param page_size: (Optional) use paging and get this many items in each page


   .. py:method:: last_updated_after(self, last_updated_after)

      :param last_updated_after: (Optional) Only get records newer than this


   .. py:method:: last_updated_before(self, last_updated_before)

      :param last_updated_before: (Optional) Only get records older than this


   .. py:method:: sort_fields(self, sort_fields)

      :param sort_fields: sort by fields in the resource


   .. py:method:: auth_server_url(self, auth_server_url)

      :param auth_server_url: server url to call to get the authentication token


   .. py:method:: auth_scopes(self, auth_scopes)

      :param auth_scopes: list of scopes to request permission for e.g., system/AllergyIntolerance.read


   .. py:method:: login_token(self, login_token)

      :param login_token: login token to use


   .. py:method:: client_credentials(self, client_id, client_secret)

      Sets client credentials to use when calling the FHIR server


      :param client_id: client_id
      :param client_secret: client_secret
      :return: self


   .. py:method:: logger(self, logger)

      Logger to use for logging calls to the FHIR server


      :param logger: logger


   .. py:method:: adapter(self, adapter)

      Http Adapter to use for calling the FHIR server


      :param adapter: adapter


   .. py:method:: limit(self, limit)

      Limit the results


      :param limit: Limit results to this count


   .. py:method:: access_token_async(self)
      :property:
      :async:

      Gets current access token


      :return: access token if any


   .. py:method:: set_access_token(self, value)

      Sets access token


      :param value: access token


   .. py:method:: delete_async(self)
      :async:

      Delete the resources


   .. py:method:: delete(self)

      Delete the resources


   .. py:method:: separate_bundle_resources(self, separate_bundle_resources)

      Set flag to separate bundle resources


      :param separate_bundle_resources:


   .. py:method:: expand_fhir_bundle(self, expand_fhir_bundle)

      Set flag to expand the FHIR bundle into a list of resources. If false then we don't un bundle the response


      :param expand_fhir_bundle: whether to just return the result as a FHIR bundle


   .. py:method:: get_async(self)
      :async:

      Issues a GET call


   .. py:method:: get(self)


   .. py:method:: create_http_session()
      :staticmethod:

      Creates an HTTP Session


   .. py:method:: get_with_handler_async(self, session, page_number, ids, fn_handle_batch, fn_handle_error, id_above = None)
      :async:


   .. py:method:: get_page_by_query_async(self, session, start_page, increment, output_queue, fn_handle_batch, fn_handle_error)
      :async:

      Gets the specified page for query
      :param session:
      :param start_page:
      :param increment:
      :param output_queue:
      :param fn_handle_batch:
      :param fn_handle_error:
      :return: list of paging results


   .. py:method:: get_page_by_query_tasks_async(self, concurrent_requests, output_queue, fn_handle_batch, fn_handle_error, http)
      :async:

      Returns tasks to get pages by query
      :param concurrent_requests:
      :param output_queue:
      :param fn_handle_batch:
      :param fn_handle_error:
      :param http:
      :return: task


   .. py:method:: get_by_query_in_pages_async(self, concurrent_requests, output_queue, fn_handle_batch, fn_handle_error)
      :async:

      Retrieves the data in batches (using paging) to reduce load on the FHIR server and to reduce network traffic

      :param output_queue:
      :type output_queue:
      :param fn_handle_error:
      :param concurrent_requests:
      :param fn_handle_batch: function to call for each batch.  Receives a list of resources where each
                                  resource is a dictionary. If this is specified then we don't return
                                  the resources anymore.  If this function returns False then we stop
                                  processing batches.
      :return response containing all the resources received


   .. py:method:: authenticate_async(self, http, auth_server_url, auth_scopes, login_token)
      :async:

      Authenticates with an OAuth Provider


      :param http: http session
      :param auth_server_url: url to auth server /token endpoint
      :param auth_scopes: list of scopes to request
      :param login_token: login token to use for authenticating
      :return: access token


   .. py:method:: merge_async(self, json_data_list)
      :async:

      Calls $merge function on FHIR server


      :param json_data_list: list of resources to send


   .. py:method:: merge(self, json_data_list)

      Calls $merge function on FHIR server


      :param json_data_list: list of resources to send


   .. py:method:: graph_async(self, *, graph_definition, contained, process_in_batches = None, fn_handle_batch = None, fn_handle_error = None, concurrent_requests = 1)
      :async:

      Executes the $graph query on the FHIR server


      :param fn_handle_error:
      :param concurrent_requests:
      :param graph_definition: definition of a graph to execute
      :param contained: whether we should return the related resources as top level list or nest them inside their
                          parent resources in a contained property
      :param process_in_batches: whether to process in batches of size page_size
      :param fn_handle_batch: Optional function to execute on each page of data.  Note that if this is passed we don't
                              return the resources in the response anymore.  If this function returns false then we
                              stop processing any further batches.


   .. py:method:: graph(self, *, graph_definition, contained, process_in_batches = None, concurrent_requests = 1)


   .. py:method:: include_total(self, include_total)

      Whether to ask the server to include the total count in the result

      :param include_total: whether to include total count


   .. py:method:: filter(self, filter_)

      Allows adding in a custom filters that derives from BaseFilter


      :param filter_: list of custom filter instances that derives from BaseFilter.


   .. py:method:: update_async(self, json_data)
      :async:

      Update the resource.  This will completely overwrite the resource.  We recommend using merge()
          instead since that does proper merging.

      :param json_data: data to update the resource with


   .. py:method:: update(self, json_data)


   .. py:method:: get_resources_by_id_in_parallel_batches_async(self, concurrent_requests, chunks, fn_handle_batch, fn_handle_error)
      :async:


   .. py:method:: get_resources_by_id_async(self, session, queue, task_number, fn_handle_batch, fn_handle_error)
      :async:


   .. py:method:: divide_into_chunks(array, chunk_size)
      :staticmethod:


   .. py:method:: handle_error(self, error, response, page_number)


   .. py:method:: get_resources_by_query_and_last_updated_async(self, last_updated_start_date, last_updated_end_date, concurrent_requests = 10, page_size_for_retrieving_resources = 100, page_size_for_retrieving_ids = 10000, fn_handle_batch = None, fn_handle_error = None)
      :async:

      Gets results for a query by paging through one day at a time,
          first downloading all the ids and then retrieving resources for each id in parallel
      :param fn_handle_error: Optional function to call when there is an error
      :param fn_handle_batch: Optional function to call when a batch is downloaded
      :param last_updated_start_date: find resources updated after this datetime
      :param last_updated_end_date: find resources updated before this datetime
      :param concurrent_requests: number of concurrent requests to make to the server
      :param page_size_for_retrieving_resources: number of resources to download in one batch
      :param page_size_for_retrieving_ids:: number of ids to download in one batch


   .. py:method:: get_resources_by_query_and_last_updated(self, last_updated_start_date, last_updated_end_date, concurrent_requests = 10, page_size_for_retrieving_resources = 100, page_size_for_retrieving_ids = 10000, fn_handle_batch = None, fn_handle_error = None)

      Gets results for a query by paging through one day at a time,
          first downloading all the ids and then retrieving resources for each id in parallel
      :param fn_handle_error: Optional function to call when there is an error
      :param fn_handle_batch: Optional function to call when a batch is downloaded
      :param last_updated_start_date: find resources updated after this datetime
      :param last_updated_end_date: find resources updated before this datetime
      :param concurrent_requests: number of concurrent requests to make to the server
      :param page_size_for_retrieving_resources: number of resources to download in one batch
      :param page_size_for_retrieving_ids:: number of ids to download in one batch


   .. py:method:: get_resources_by_query_async(self, last_updated_start_date = None, last_updated_end_date = None, concurrent_requests = 10, page_size_for_retrieving_resources = 100, page_size_for_retrieving_ids = 10000, fn_handle_batch = None, fn_handle_error = None)
      :async:

      Gets results for a query by first downloading all the ids and then retrieving resources for each id in parallel
      :param fn_handle_error: Optional function to call when there is an error
      :param fn_handle_batch: Optional function to call when a batch is downloaded
      :param last_updated_start_date: find resources updated after this datetime
      :param last_updated_end_date: find resources updated before this datetime
      :param concurrent_requests: number of concurrent requests to make to the server
      :param page_size_for_retrieving_resources: number of resources to download in one batch
      :param page_size_for_retrieving_ids:: number of ids to download in one batch


   .. py:method:: get_ids_for_query_async(self, *, last_updated_start_date = None, last_updated_end_date = None, concurrent_requests = 10, page_size_for_retrieving_ids = 10000)
      :async:

      Gets just the ids of the resources matching the query


      :param last_updated_start_date:
      :param last_updated_end_date:
      :param concurrent_requests:
      :param page_size_for_retrieving_ids:
      :return: list of ids


   .. py:method:: get_ids_for_query(self, *, last_updated_start_date = None, last_updated_end_date = None, concurrent_requests = 10, page_size_for_retrieving_ids = 10000)

      Gets just the ids of the resources matching the query


      :param last_updated_start_date:
      :param last_updated_end_date:
      :param concurrent_requests:
      :param page_size_for_retrieving_ids:
      :return: list of ids


   .. py:method:: get_resources_by_query(self, last_updated_start_date = None, last_updated_end_date = None, concurrent_requests = 10, page_size_for_retrieving_resources = 100, page_size_for_retrieving_ids = 10000, fn_handle_batch = None, fn_handle_error = None)

      Gets results for a query by first downloading all the ids and then retrieving resources for each id in parallel
      :param fn_handle_error: Optional function to call when there is an error
      :param fn_handle_batch: Optional function to call when a batch is downloaded
      :param last_updated_start_date: find resources updated after this datetime
      :param last_updated_end_date: find resources updated before this datetime
      :param concurrent_requests: number of concurrent requests to make to the server
      :param page_size_for_retrieving_resources: number of resources to download in one batch
      :param page_size_for_retrieving_ids:: number of ids to download in one batch


   .. py:method:: get_in_batches_async(self, fn_handle_batch)
      :async:

      Retrieves the data in batches (using paging) to reduce load on the FHIR server and to reduce network traffic

      :param fn_handle_batch: function to call for each batch.  Receives a list of resources where each
                                  resource is a dictionary. If this is specified then we don't return
                                  the resources anymore.  If this function returns False then we stop
                                  processing batches.
      :return response containing all the resources received


   .. py:method:: get_in_batches(self, fn_handle_batch = None)

      Retrieves the data in batches (using paging) to reduce load on the FHIR server and to reduce network traffic

      :param fn_handle_batch: function to call for each batch.  Receives a list of resources where each
                                  resource is a dictionary. If this is specified then we don't return
                                  the resources anymore.  If this function returns False then we stop
                                  processing batches.
      :return response containing all the resources received



