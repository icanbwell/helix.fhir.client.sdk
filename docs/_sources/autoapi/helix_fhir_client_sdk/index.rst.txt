:py:mod:`helix_fhir_client_sdk`
===============================

.. py:module:: helix_fhir_client_sdk


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   exceptions/index.rst
   filters/index.rst
   graph/index.rst
   loggers/index.rst
   responses/index.rst
   validators/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   fhir_client/index.rst
   well_known_configuration/index.rst


