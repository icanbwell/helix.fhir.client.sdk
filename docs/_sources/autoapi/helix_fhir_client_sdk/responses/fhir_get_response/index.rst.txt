:py:mod:`helix_fhir_client_sdk.responses.fhir_get_response`
===========================================================

.. py:module:: helix_fhir_client_sdk.responses.fhir_get_response


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse




.. py:class:: FhirGetResponse(url, responses, error, access_token)

   
   Class that encapsulates the response from FHIR server

   :param url: url that was being accessed
   :param responses: response text
   :param error: Any error returned by FHIR server
   :param access_token: access token that was used


