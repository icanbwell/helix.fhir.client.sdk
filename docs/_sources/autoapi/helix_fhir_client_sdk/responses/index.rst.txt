:py:mod:`helix_fhir_client_sdk.responses`
=========================================

.. py:module:: helix_fhir_client_sdk.responses


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   fhir_get_response/index.rst
   fhir_merge_response/index.rst


