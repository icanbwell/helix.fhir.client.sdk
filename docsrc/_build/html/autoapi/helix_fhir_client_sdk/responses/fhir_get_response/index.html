

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>helix_fhir_client_sdk.responses.fhir_get_response &mdash; helix.fhir.client.sdk  documentation</title>
  

  
  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/graphviz.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/underscore.js"></script>
        <script src="../../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../../index.html" class="icon icon-home"> helix.fhir.client.sdk
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">API Reference</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">helix.fhir.client.sdk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><code class="xref py py-mod docutils literal notranslate"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response</span></code></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../../_sources/autoapi/helix_fhir_client_sdk/responses/fhir_get_response/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-helix_fhir_client_sdk.responses.fhir_get_response">
<span id="helix-fhir-client-sdk-responses-fhir-get-response"></span><h1><a class="reference internal" href="#module-helix_fhir_client_sdk.responses.fhir_get_response" title="helix_fhir_client_sdk.responses.fhir_get_response"><code class="xref py py-mod docutils literal notranslate"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response</span></code></a><a class="headerlink" href="#module-helix_fhir_client_sdk.responses.fhir_get_response" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<div class="section" id="classes">
<h3>Classes<a class="headerlink" href="#classes" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse" title="helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FhirGetResponse</span></code></a></p></td>
<td><p>Class that encapsulates the response from FHIR server</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response.</span></span><span class="sig-name descname"><span class="pre">FhirGetResponse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">responses</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">access_token</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse" title="Permalink to this definition">¶</a></dt>
<dd><p>Class that encapsulates the response from FHIR server</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url</strong> (<em>str</em>) – url that was being accessed</p></li>
<li><p><strong>responses</strong> (<em>str</em>) – response text</p></li>
<li><p><strong>error</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Any error returned by FHIR server</p></li>
<li><p><strong>access_token</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – access token that was used</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Helix Team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>