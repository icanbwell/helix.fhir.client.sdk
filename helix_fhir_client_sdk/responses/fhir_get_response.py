from typing import Optional


class FhirGetResponse:
    def __init__(
        self,
        url: str,
        responses: str,
        error: Optional[str],
        access_token: Optional[str],
        total_count: Optional[int],
        status: int,
    ) -> None:
        """
        Class that encapsulates the response from FHIR server

        :param url: url that was being accessed
        :param responses: response text
        :param error: Any error returned by FHIR server
        :param access_token: access token that was used
        :param total_count: count of total records that match the provided query.
                            Only set if include_total_count was set to avoid expensive operation by server.
        """
        self.url: str = url
        self.responses: str = responses
        self.error: Optional[str] = error
        self.access_token: Optional[str] = access_token
        self.total_count: Optional[int] = total_count
        self.status: int = status
