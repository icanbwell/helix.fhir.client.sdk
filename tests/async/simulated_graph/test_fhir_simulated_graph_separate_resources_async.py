import json
from pathlib import Path
from typing import Dict, Any

from mockserver_client.mockserver_client import (
    MockServerFriendlyClient,
    mock_request,
    mock_response,
    times,
)

from helix_fhir_client_sdk.fhir_client import FhirClient

from helix_fhir_client_sdk.responses.fhir_get_response import FhirGetResponse


async def test_fhir_simulated_graph_async() -> None:
    print("")
    data_dir: Path = Path(__file__).parent.joinpath("./")
    graph_json: Dict[str, Any]
    with open(data_dir.joinpath("graphs").joinpath("aetna.json"), "r") as file:
        contents = file.read()
        graph_json = json.loads(contents)

    test_name = "test_fhir_simulated_graph_async"

    mock_server_url = "http://mock-server:1080"
    mock_client: MockServerFriendlyClient = MockServerFriendlyClient(
        base_url=mock_server_url
    )

    relative_url: str = test_name
    absolute_url: str = mock_server_url + "/" + test_name

    mock_client.clear(f"/{test_name}/*.*")
    mock_client.reset()

    response_text = {
        "resourceType": "Patient",
        "id": "1",
        "generalPractitioner": [{"reference": "Practitioner/5"}],
        "managingOrganization": {"reference": "Organization/6"},
    }

    mock_client.expect(
        mock_request(path=f"/{relative_url}/Patient/1", method="GET"),
        mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Practitioner", "id": "5"}
    mock_client.expect(
        mock_request(path=f"/{relative_url}/Practitioner/5", method="GET"),
        mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Organization", "id": "6"}
    mock_client.expect(
        mock_request(path=f"/{relative_url}/Organization/6", method="GET"),
        mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {
        "entry": [
            {
                "resource": {
                    "resourceType": "Coverage",
                    "id": "7",
                    "payor": [{"reference": "Organization/CoveragePayor"}],
                }
            }
        ]
    }
    mock_client.expect(
        mock_request(
            path=f"/{relative_url}/Coverage",
            querystring={"patient": "1"},
            method="GET",
        ),
        mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Organization", "id": "CoveragePayor"}
    mock_client.expect(
        mock_request(path=f"/{relative_url}/Organization/CoveragePayor", method="GET"),
        mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {
        "entry": [{"resource": {"resourceType": "Observation", "id": "8"}}]
    }
    mock_client.expect(
        mock_request(
            path=f"/{relative_url}/Observation",
            querystring={
                "patient": "1",
                "category": "vital-signs,social-history,laboratory",
            },
            method="GET",
        ),
        mock_response(body=response_text),
        timing=times(1),
    )

    fhir_client = FhirClient()

    fhir_client = fhir_client.log_level("DEBUG")

    fhir_client = fhir_client.url(absolute_url).resource("Patient")

    fhir_client.extra_context_to_return({"service_slug": "medstar"})

    auth_access_token = "my_access_token"
    if auth_access_token:
        fhir_client = fhir_client.set_access_token(auth_access_token)

    response: FhirGetResponse = await fhir_client.simulate_graph_async(
        id_="1",
        graph_json=graph_json,
        contained=False,
        separate_bundle_resources=True,
    )
    print(response.responses)

    expected_json = {
        "Patient": [
            {
                "resourceType": "Patient",
                "id": "1",
                "generalPractitioner": [{"reference": "Practitioner/5"}],
                "managingOrganization": {"reference": "Organization/6"},
            }
        ],
        "Practitioner": [{"resourceType": "Practitioner", "id": "5"}],
        "Organization": [
            {"resourceType": "Organization", "id": "6"},
            {"resourceType": "Organization", "id": "CoveragePayor"},
        ],
        "Coverage": [
            {
                "resourceType": "Coverage",
                "id": "7",
                "payor": [{"reference": "Organization/CoveragePayor"}],
            }
        ],
        "OperationOutcome": [
            {
                "resourceType": "OperationOutcome",
                "issue": [
                    {
                        "severity": "error",
                        "code": "not-found",
                        "details": {
                            "coding": [
                                {
                                    "system": "https://www.icanbwell.com/url",
                                    "code": "http://mock-server:1080/test_fhir_simulated_graph_async/ExplanationOfBenefit?patient=1",
                                },
                                {
                                    "system": "https://www.icanbwell.com/resourceType",
                                    "code": "ExplanationOfBenefit",
                                },
                                {
                                    "system": "https://www.icanbwell.com/statuscode",
                                    "code": 404,
                                },
                                {
                                    "system": "https://www.icanbwell.com/accessToken",
                                    "code": "my_access_token",
                                },
                            ]
                        },
                        "diagnostics": '{"url": "http://mock-server:1080/test_fhir_simulated_graph_async/ExplanationOfBenefit?patient=1", "error": "NotFound", "status": 404, "extra_context_to_return": {"service_slug": "medstar"}, "accessToken": "my_access_token", "requestId": null, "resourceType": "ExplanationOfBenefit", "id": null}',
                    }
                ],
            },
            {
                "resourceType": "OperationOutcome",
                "issue": [
                    {
                        "severity": "error",
                        "code": "not-found",
                        "details": {
                            "coding": [
                                {
                                    "system": "https://www.icanbwell.com/url",
                                    "code": "http://mock-server:1080/test_fhir_simulated_graph_async/MedicationRequest?patient=1",
                                },
                                {
                                    "system": "https://www.icanbwell.com/resourceType",
                                    "code": "MedicationRequest",
                                },
                                {
                                    "system": "https://www.icanbwell.com/statuscode",
                                    "code": 404,
                                },
                                {
                                    "system": "https://www.icanbwell.com/accessToken",
                                    "code": "my_access_token",
                                },
                            ]
                        },
                        "diagnostics": '{"url": "http://mock-server:1080/test_fhir_simulated_graph_async/MedicationRequest?patient=1", "error": "NotFound", "status": 404, "extra_context_to_return": {"service_slug": "medstar"}, "accessToken": "my_access_token", "requestId": null, "resourceType": "MedicationRequest", "id": null}',
                    }
                ],
            },
            {
                "resourceType": "OperationOutcome",
                "issue": [
                    {
                        "severity": "error",
                        "code": "not-found",
                        "details": {
                            "coding": [
                                {
                                    "system": "https://www.icanbwell.com/url",
                                    "code": "http://mock-server:1080/test_fhir_simulated_graph_async/MedicationDispense?patient=1",
                                },
                                {
                                    "system": "https://www.icanbwell.com/resourceType",
                                    "code": "MedicationDispense",
                                },
                                {
                                    "system": "https://www.icanbwell.com/statuscode",
                                    "code": 404,
                                },
                                {
                                    "system": "https://www.icanbwell.com/accessToken",
                                    "code": "my_access_token",
                                },
                            ]
                        },
                        "diagnostics": '{"url": "http://mock-server:1080/test_fhir_simulated_graph_async/MedicationDispense?patient=1", "error": "NotFound", "status": 404, "extra_context_to_return": {"service_slug": "medstar"}, "accessToken": "my_access_token", "requestId": null, "resourceType": "MedicationDispense", "id": null}',
                    }
                ],
            },
        ],
        "Observation": [{"resourceType": "Observation", "id": "8"}],
    }

    assert json.loads(response.responses) == expected_json
