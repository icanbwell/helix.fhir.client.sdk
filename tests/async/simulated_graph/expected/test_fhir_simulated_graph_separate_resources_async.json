{
  "Patient": [
    {
      "resourceType": "Patient",
      "id": "1",
      "generalPractitioner": [
        {
          "reference": "Practitioner/5"
        }
      ],
      "managingOrganization": {
        "reference": "Organization/6"
      }
    }
  ],
  "Practitioner": [
    {
      "resourceType": "Practitioner",
      "id": "5"
    }
  ],
  "Organization": [
    {
      "resourceType": "Organization",
      "id": "6"
    },
    {
      "resourceType": "Organization",
      "id": "CoveragePayor"
    }
  ],
  "Coverage": [
    {
      "resourceType": "Coverage",
      "id": "7",
      "payor": [
        {
          "reference": "Organization/CoveragePayor"
        }
      ]
    }
  ],
  "OperationOutcome": [
    {
      "resourceType": "OperationOutcome",
      "issue": [
        {
          "severity": "error",
          "code": "not-found",
          "details": {
            "coding": [
              {
                "system": "https://www.icanbwell.com/url",
                "code": "http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/ExplanationOfBenefit?patient=1"
              },
              {
                "system": "https://www.icanbwell.com/resourceType",
                "code": "ExplanationOfBenefit"
              },
              {
                "system": "https://www.icanbwell.com/statuscode",
                "code": 404
              },
              {
                "system": "https://www.icanbwell.com/accessToken",
                "code": "my_access_token"
              }
            ]
          },
          "diagnostics": "{\"url\": \"http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/ExplanationOfBenefit?patient=1\", \"error\": \"NotFound\", \"status\": 404, \"extra_context_to_return\": {\"service_slug\": \"medstar\"}, \"accessToken\": \"my_access_token\", \"requestId\": null, \"resourceType\": \"ExplanationOfBenefit\", \"id\": null}"
        }
      ]
    },
    {
      "resourceType": "OperationOutcome",
      "issue": [
        {
          "severity": "error",
          "code": "not-found",
          "details": {
            "coding": [
              {
                "system": "https://www.icanbwell.com/url",
                "code": "http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/MedicationRequest?patient=1"
              },
              {
                "system": "https://www.icanbwell.com/resourceType",
                "code": "MedicationRequest"
              },
              {
                "system": "https://www.icanbwell.com/statuscode",
                "code": 404
              },
              {
                "system": "https://www.icanbwell.com/accessToken",
                "code": "my_access_token"
              }
            ]
          },
          "diagnostics": "{\"url\": \"http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/MedicationRequest?patient=1\", \"error\": \"NotFound\", \"status\": 404, \"extra_context_to_return\": {\"service_slug\": \"medstar\"}, \"accessToken\": \"my_access_token\", \"requestId\": null, \"resourceType\": \"MedicationRequest\", \"id\": null}"
        }
      ]
    },
    {
      "resourceType": "OperationOutcome",
      "issue": [
        {
          "severity": "error",
          "code": "not-found",
          "details": {
            "coding": [
              {
                "system": "https://www.icanbwell.com/url",
                "code": "http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/MedicationDispense?patient=1"
              },
              {
                "system": "https://www.icanbwell.com/resourceType",
                "code": "MedicationDispense"
              },
              {
                "system": "https://www.icanbwell.com/statuscode",
                "code": 404
              },
              {
                "system": "https://www.icanbwell.com/accessToken",
                "code": "my_access_token"
              }
            ]
          },
          "diagnostics": "{\"url\": \"http://mock-server:1080/test_fhir_simulated_graph_separate_resources_async/MedicationDispense?patient=1\", \"error\": \"NotFound\", \"status\": 404, \"extra_context_to_return\": {\"service_slug\": \"medstar\"}, \"accessToken\": \"my_access_token\", \"requestId\": null, \"resourceType\": \"MedicationDispense\", \"id\": null}"
        }
      ]
    }
  ],
  "Observation": [
    {
      "resourceType": "Observation",
      "id": "8"
    }
  ]
}
